### PR TITLE
Add the reference predicate

### PR DESCRIPTION
**Pull request by mdeicas**
Adds the specification of a predicate to address #179.

One question I had was how to express what content the reference contains. I think the current use cases are met by the MIME type, and adding another field to specify this more explicitly would make the predicate more complex. If needed, it can be added in the future in a new version.

Open to feedback!

### CODE REVIEW

**Review by pxp928 — APPROVED** [review]
This looks good to me for a first revision Thank You!

File: spec/predicates/reference.md
Comment (TomHennen) [inline spec/predicates/reference.md:74]: Should `downloadLocation` be "MUST"?  How useful would a reference attestation that doesn't point to something be?

Comment (TomHennen) [inline spec/predicates/reference.md:74]: optional: should `mediaType` be required with a suggested 'default' type if the user doesn't know what it is?

File: spec/predicates/reference.md
Comment (TomHennen) [inline spec/predicates/reference.md:24]: Can you say more about why SCAI isn't quite right for this case?  (E.g. can we answer [this question](https://github.com/in-toto/attestation/blob/main/docs/new_predicate_guidelines.md#:~:text=Why%20don%E2%80%99t%20existing%20predicates%20cover%20this%20use%20case%3F))

**Comment by TomHennen** [comment]
Thanks for sending this!

Would it be possible to create a [protobuf definition](https://github.com/in-toto/attestation/blob/main/docs/new_predicate_guidelines.md#:~:text=To%20generate%20Go/Python/Java%20language%20bindings%20for%20the%20new%20predicate%2C%20include%20a%20protobuf%20definition.)

File: spec/predicates/reference.md
Comment (mdeicas) [reply spec/predicates/reference.md]: Makes sense to me, changed to MUST.

Comment (mdeicas) [reply spec/predicates/reference.md]: `application/octet-stream` [is specified](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types#:~:text=contain%20binary%20data.-,application/octet%2Dstream,-is%20the%20default) as the default MIME type for unknown file types. Having a required field is more usable, so this sounds good to me.

File: spec/predicates/reference.md
Comment (mdeicas) [reply spec/predicates/reference.md]: Done.

**Comment by marcelamelara** [comment]
I don't see how SCAI can't cover this use case. In fact, we have a [few](https://github.com/in-toto/scai-demos/tree/main/examples/sbom%2Bslsa) [examples]((https://github.com/in-toto/attestation/blob/main/spec/predicates/scai.md#attestation-for-evidence-collection) that show how SCAI can be used to transmit SBOM (or any other type of reference to another doc) without having to embed an entire SBOM in a Statement. 

Rather than re-inventing the wheel for one specific use case, I'd much rather we do attempt to standardize around specific attribute values (e.g., our example in scai-demos uses the GUAC verbs `HasSBOM` and `HasSLSA`), and declare certain fields in the SCAI predicate as required when used in certain ways, not unlike how we do for [DigestSets](https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md#supported-algorithms), where the spec does in fact standardize certain field values.

**Comment by marcelamelara** [comment]
To add, I'm absolutely happy to work with you to explore different fields/values that could be standardized for this common use case (i.e. exchanging SBOM, having it consumed by GUAC). For instance, for this use case, the `target` field of the SCAI predicate may be required if `attribute: HasSBOM` and `mediaType: application/spdx+json`. Again, there's nothing in the existing in-toto attestation spec that precludes these sorts of restrictions for specific use cases.

**Comment by lumjjb** [comment]
(copying my response from the original issue since it may be relevant to discussion)

> 
> I think this is slightly different framing from the witness/provenance materials, and resembling more of an intoto link predicate (although the link predicate is a bit generic, so having a defined predicate will help the consumption/adoption by being opinionated on how fields should be used).
> 
> i can see where folks are coming from and indeed there is "encodability" overlap with the different predicates, but I agree with sentiment that the intent here is different. The fields required for supporting documents vs source code attestations vs security statements are different, and i believe it is helpful to be intentional with defining each of them. Else, we may see issues around there being too many optional fields (which also have different use based on context), and that can cause confusion. I'll use SPDX as an example (since i am a contributor and thus an offender), where a "Package" can refer to many things (sources, application bundles) and has many optional fields depending on what it represents (which 3.0 does try and solve by providing a bit more specificity in subclassing, but still suffers from legacy convention).

Agreed on the overall goal of having a "predicate dictionary" type construct where we define here are all the predicates or properties and having guidance on how to consume them. It is difficult for me (who is perhaps uninformed) to understand the delineation around predicates and when something is a separate enough usecase for its own predicate. e.g. SCAI vs vulnerability, intoto link vs SLSA.

**Comment by mdeicas** [comment]
Thanks for the responses. IMO a new predicate type has a few benefits over SCAI, but I agree that SCAI does work for the current use case at a technical level, and I'm happy to go with the option that the maintainers think is best. Is the SCAI predicate type spec a good place to define the attribute?

The benefits I see for this dedicated predicate type is that it can evolve more easily (e.g. adding new fields) and is simpler to use for consumers (possibly fewer changes to the spec and the data "type" is encoded at the statement layer instead of the predicate layer).

**Comment by marcelamelara** [comment]
Sorry for my delay.

> It is difficult for me (who is perhaps uninformed) to understand the delineation around predicates and when something is a separate enough usecase for its own predicate.

This is totally fair, and the more general feedback for us here is that we should set clearer guidelines for users of in-toto attestations to choose the right predicate(s), or describe more clearly when it really makes sense to create new ones. We have an upcoming maintainer's meeting, this would be a good topic to raise. Thanks!!

> Is the SCAI predicate type spec a good place to define the attribute?

Yes, what I'm imagining is adding a new section to the spec titled something like "Supported Assertions" where we would catalogue the GUAC attributes and expected other field requirements/values for those assertions. This would also entail adding the appropriate verification logic to the SCAI APIs, much how we validate [supported hash algorithms](https://github.com/in-toto/attestation/blob/main/go/v1/resource_descriptor.go) in resource descriptors.

>The benefits I see for this dedicated predicate type is that it can evolve more easily (e.g. adding new fields) and is simpler to use for consumers (possibly fewer changes to the spec and the data "type" is encoded at the statement layer instead of the predicate layer).

I definitely understand this rationale. And the predicates are meant to be living specs that evolve as the community gathers feedback from users/consumers about use cases or usability. This is meant to be a community-driven process, so subsequent versions of the SCAI predicate (or any other) can definitely be amended to add/change fields, and better suit the broad use cases they were designed for.

Could you please explain what you mean by "the type is encoded at the statement layer" in the case of the dedicated reference predicate? If a consumer just sees `predicateType: "https://in-toto.io/attestation/reference/v0.1"`, it still needs to step into the predicate layer and iterate over the list of references in order to understand what types of documents/resources are being referenced. I could just be missing something.

**Review by TomHennen — APPROVED** [review]

**Comment by marcelamelara** [comment]
Given that we have the needed approvals for this PR, I won't stand in the way of merging.

File: spec/predicates/reference.md
Comment (mdeicas) [reply spec/predicates/reference.md]: I think if the the `uri` is left empty and `digest` is provided, then wouldn't be a conflict.

**Pull request merged**
